src/ServiceControl.LearningTransport/QueueLengthProvider.cs

```diff
@@ -0,0 +1,121 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Logging;
+    using Metrics;
+    using ServiceControl.Monitoring.Infrastructure;
+    using ServiceControl.Monitoring.Messaging;
+    using ServiceControl.Monitoring.QueueLength;
+
+    public class QueueLengthProvider : IProvideQueueLength
+    {
+        string rootFolder;
+        QueueLengthStore queueLengthStore;
+        ConcurrentDictionary<EndpointInputQueue, EndpointInputQueue> endpointsHash = new ConcurrentDictionary<EndpointInputQueue, EndpointInputQueue>();
+        CancellationTokenSource cancel;
+        Task task;
+
+        static TimeSpan QueryDelayInterval = TimeSpan.FromMilliseconds(200);
+
+        public void Initialize(string connectionString, QueueLengthStore store)
+        {
+            rootFolder = Path.Combine(connectionString, ".learningtransport");
+            queueLengthStore = store;
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+        {
+            var key = new EndpointInputQueue(endpointInstanceId.EndpointName, metadataReport.LocalAddress);
+            endpointsHash.AddOrUpdate(key, key, (_, __) => key);
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, TaggedLongValueOccurrence metricsReport)
+        {
+            // The endpoint should not be sending this data
+        }
+
+        public Task Start()
+        {
+            cancel = new CancellationTokenSource();
+            task = Task.Run(async () =>
+            {
+                while (!cancel.IsCancellationRequested)
+                {
+                    try
+                    {
+                        var queueLengths = GetQueueLengths();
+                        UpdateStore(queueLengths);
+
+                        await Task.Delay(QueryDelayInterval, cancel.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // It's OK. We're shutting down
+                    }
+                    catch (Exception ex)
+                    {
+                        Log.Warn("Problem getting learning transport queue length", ex);
+                    }
+                }
+            });
+
+            return Task.FromResult(0);
+        }
+
+        void UpdateStore(ILookup<string, long?> queueLengths)
+        {
+            var now = DateTime.UtcNow.Ticks;
+            foreach (var instance in endpointsHash.Keys)
```

Comment:
Accessing keys, locks the dictionary. Please use enumerator over the dictionary and get instance from `kvp.Key`